[Docs] End User Doc Page (#18055)
gravitational/teleport: docs/pages/user-guides.mdx

---
title: User Guides
description: Documentation for end users
h1: End User Documentation
---

> Teleport is a complex security solution, and most of our documentation covers

Reviewer: I don't think we need this first sentence.

Author: Removed and added more context.

the installation, configuration, and management of Teleport for administrators.

> For end users, this page covers the basics of authenticating to Teleport and

Reviewer: ```suggestion
This guide covers the basics of authenticating to Teleport and
```

accessing resources.

> ## tsh

Reviewer: What would you think of making this H2 about "Client tools", rather than the first two H2s about client tools and the next few about protocols? To make the structure clearer, we could also have one H2 about tools and another H2 about protocols, with each specific protocol having its own H3.

Author: I like that idea conceptually. I went back and forth on nesting headers similar to how you suggest when writing. The factor that made me go this way was that our ToC doesn't render H3s, and I wanted new end users to be able to easily see the thing they were looking for there.

I don't suppose we have a frontmatter value or something that lets us adjust the depth of ToC'd headers, is there?

Author: Adjusted headers so that the top-level ToC is broken down into "Clients" and "Protocols"

> `tsh` lets you authenticate to Teleport, list services, and initate either

Reviewer: Instead of talking about local proxies and direct connections, I think we can just say something like "connect to resources". We don't mention how to use a local proxy until later, so I think we can simplify this here.

Author: Adjusting

Tap any security key
> Profile URL:        https://teleport.example.com:443
  Logged in as:       alice
  Cluster:            example.com
  Roles:              access
  Logins:             root, ubuntu, ec2-user
  Kubernetes:         enabled
  Valid until:        2022-11-01 22:37:05 -0500 CDT [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty, private-key-policy
```

</ScopedBlock>

<ScopedBlock scope={["cloud"]}>

```code
$ tsh login --proxy=mytenant.teleport.sh --user=alice
Enter password for Teleport user alice:
Tap any security key
> Profile URL:        https://teleport.example.com:443
  Logged in as:       alice
  Cluster:            mytennant.teleport.sh
  Roles:              access
  Logins:             root, ubuntu, ec2-user
  Kubernetes:         enabled
  Valid until:        2022-11-01 22:37:05 -0500 CDT [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty, private-key-policy
```

</ScopedBlock>

Depending on how Teleport was configured for your network, you may need
additional flags, such as `--auth=a_saml_connector`. Your administrator should

> provde the details required for your particular use case.

Reviewer: ```suggestion
provide the details required for your particular use case.
```

## Teleport Connect

The Teleport Connect app provides all the same access to resources as `tsh` and

> more, in a friendly graphic user interface. After

Reviewer: Should we specify "more"?

Author: I actually don't think there is "more" in this context, so I'm removing it.

[downloading](https://goteleport.com/download/) and installing Teleport Connect,

> you can log in initiate sessions for Server and Database Access all, within a

Reviewer: ```suggestion
you can log in and initiate sessions for Server and Database Access within a
```

single window.

1. Click **CONNECT** to connect to the Teleport cluster:

   ![A new instance of Teleport Connect](../img/teleport-connect-init.png)

2. Provide the address of your Teleport Cluster (e.g. 
<ScopedBlock scope={["oss", "enterprise"]}>`https://teleport.example.com`</ScopedBlock>
<ScopedBlock scope={["cloud"]}>`https://mytennant.teleport.sh`</ScopedBlock>
) and click **NEXT**.

3. Teleport Connect will ask you for your username, password, and MFA. If this

> Teleport plane is integrated with an external Identity Provider (**IdP**), it

Reviewer: "If this Teleport plane" -> "If Teleport"

may instead prompt you to authenticate in a browser window with that service.

4. Once authenticated, you can browse and connect to all the resources your
user is permitted to access:

   ![An example of Teleport Connect, populated with instances](../img/use-teleport/connect-ui-overview.png)

> 

Reviewer: Should we have a section for the Web UI?

> ## Server Access

Reviewer: As mentioned above, I'd be in favor of having one H2 for tools and another H2 for protocols so we have a clearer organizational structure

> <Tabs>

Reviewer: I think there should be a `TabItem` for the Web UI here and in other protocol-specific sections.

<TabItem label="tsh">

> `tsh ls` lists the servers you have access to through Server Access (ssh):

Reviewer: ```suggestion
`tsh ls` lists the servers you have access to through Server Access (SSH):
```

```code
$ tsh ls
Node Name           Address         Labels                       
------------------- --------------- ---------------------------- 
server1.example.com 192.0.2.24:3022 access=servers,hostname=server1
server2.example.com 192.0.2.32:3022 access=servers,hostname=server2
```

To connect to a server:

```code
$ tsh ssh ubuntu@server1.example.com
ubuntu@server1:~$
```

</TabItem>
<TabItem label="Teleport Connect">

Under the **Servers** tab, click **CONNECT** next to the server you want to
access. Select or type in a username available to your Teleport user.

In a new tab, Teleport Connect will initiate the connection and provide a terminal environment.

</TabItem>
</Tabs>

> ## Transferring files

Reviewer: Does this need its own H2? Should we add it to the `tsh` section instead?

Author: Relating to my ToC comment above, I was thinking about headers in terms of readers being able to scan them for the thing they are looking to do at that moment, and be brought directly to the needful. But I take your point.

`tsh scp` will let you transfer files to servers behind Teleport:

```code
$ tsh scp somefile.ext xeon.tp.alexf.xyz:
somefile.ext   7% |███████                | (25/342 MB, 2.9 MB/s) [9s:1m48s]
```

## Database Access

<Tabs>
<TabItem label="tsh">

`tsh db ls` will list the databases available to your user:

```code
$ tsh db ls
Name                                               Description Allowed Users   Labels  Connect
-----------------------------------------------    ----------- --------------- ------- -------------------------
myelastic                                                      [alice elastic] env=dev
mysql-server1 (user: alice, db: teleport_example)              [alice elastic] env=dev tsh db connect mysql-server1
```

> To connect to a database server, through `tsh`, you'll need a local client for

Reviewer: ```suggestion
To connect to a database server through `tsh`, you'll need a local client for
```

that database. For example, to connect to a MySQL or MariaDB database, you'll
need the [MySQL CLI](https://dev.mysql.com/doc/refman/8.0/en/mysql.html) client:

```code

> tsh db connect --db-user=alice --db-name=teleport_example mysql-server1 

Reviewer: ```suggestion
$ tsh db connect --db-user=alice --db-name=teleport_example mysql-server1 
```

+--------------------+
4 rows in set (0.09 sec)
```

For other connection types, or to connect a graphic database client, you can create a local tunnel to point your software to:

```code
$ tsh proxy db myelastic --db-user=alice --tunnel
Started authenticated tunnel for the Elasticsearch database "myelastic" in cluster "example.come" on 192.0.2.58:52669.

Use one of the following commands to connect to the database:

  * interactive SQL connection:

  $ elasticsearch-sql-cli http://localhost:52669/

  * run single request with curl:

  $ curl http://localhost:52669/

Database certificate renewed: valid until 2022-11-02T06:11:50Z [valid for 9h55m0s]
```

</TabItem>
<TabItem label="Teleport Connect">

In the **Databases** tab, select the database you want to access and click
**CONNECT**. Select or type in a username available to your Teleport user.

Teleport Connect will open a tunneled connection to the database. You can copy the port number and connect with a GUI app at `localhost`, or click **RUN** to initiate a CLI connection inside Teleport Connect:

![Teleport Connect database connection](../img/teleport-connect-mysql.png)

</TabItem>
</Tabs>

## Desktop Access

> Desktop Access is available through the Teleport web UI. In your browser,

Reviewer: ```suggestion
Desktop Access is available through the Teleport Web UI. In your browser,
```

> navigate to your Telport cluster (e.g. 

Reviewer: ```suggestion
navigate to your Teleport cluster (e.g. 
```

<ScopedBlock scope={["oss", "enterprise"]}>`https://teleport.example.com`</ScopedBlock>
<ScopedBlock scope={["cloud"]}>`https://mytennant.teleport.sh`</ScopedBlock>

> ). From the menu on the right, select <Icon name="desktop" inline size="sm"/> **Desktops**. Next to the desktop you want to access, click **CONNECT**. Select

Reviewer: ```suggestion
). 

From the menu on the right, select <Icon name="desktop" inline size="sm"/> **Desktops**. Next to the desktop you want to access, click **CONNECT**. Select
```
To break up this paragraph, which for me is on the long side.

Author: Broken up.

or type in a username available to your Teleport user. Teleport will open a new
browser tab or window and begin the RDP session. Note that you may need to wait
a moment for Teleport to log you in as the specified user.

> ## Next Steps

Reviewer: ```suggestion
## Next steps
```
We're using sentence case in docs headings

Author: Sorry, that's a hard habit to break 😄

This page is a basic introduction to using Teleport. More detailed information

> on the subject matter covered here can be found in the **Connect your Client**

Reviewer: If we move this guide into the "Connect your Client" section, I think we can remove this sentence

Author: Rephrased

section of our docs:

> - `tsh` is the primary CLI tool for Server Access through Teleport. With it,

Reviewer: ```suggestion
- `tsh` is the CLI tool for accessing resources through Teleport. With it,
```

---
title: User Guides
description: Documentation for end users
h1: End User Documentation
---

Teleport is a complex security solution, and most of our documentation covers
the installation, configuration, and management of Teleport for administrators.
For end users, this page covers the basics of authenticating to Teleport and
accessing resources.

## tsh

`tsh` lets you authenticate to Teleport, list services, and initate either
direct connections or proxy tunnels to them. After [downloading](/downloads)

> and installing `tsh`, sign in to your Teleport plane:

Reviewer 2: ```suggestion
and installing `tsh`, sign in to your Teleport cluster:
```

We don't really use the term `Teleport plane`.

<ScopedBlock scope={["oss", "enterprise"]}>

```code

> $ tsh login --proxy=teleport.example.com --user=alice

Reviewer 3: All enterprise customers will use a SAML or OIDC connector. I would use one here.

Author: Requesting input from @ptgott and @rsyracuse: It may be that I'm being overly pedantic about this, if so please check me at the door.

Author: Discussed with the team. I'm making `oss` and `enterprise` discrete scoped blocks and adding tabs in the Enterprise block for local and IdP users.

Enter password for Teleport user alice:
Tap any security key
> Profile URL:        https://teleport.example.com:443
  Logged in as:       alice
  Cluster:            example.com
  Roles:              access
  Logins:             root, ubuntu, ec2-user
  Kubernetes:         enabled
  Valid until:        2022-11-01 22:37:05 -0500 CDT [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty, private-key-policy
```

</ScopedBlock>

<ScopedBlock scope={["cloud"]}>

```code

> $ tsh login --proxy=mytennant.teleport.sh --user=alice

Reviewer 2: ```suggestion
$ tsh login --proxy=mytenant.teleport.sh --user=alice
```

Enter password for Teleport user alice:
Tap any security key
> Profile URL:        https://teleport.example.com:443
  Logged in as:       alice
  Cluster:            mytennant.teleport.sh
  Roles:              access
  Logins:             root, ubuntu, ec2-user
  Kubernetes:         enabled
  Valid until:        2022-11-01 22:37:05 -0500 CDT [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty, private-key-policy
```

</ScopedBlock>

Depending on how Teleport was configured for your network, you may need
additional flags, such as `--auth=a_saml_connector`. Your administrator should
provde the details required for your particular use case.

## Teleport Connect

The Teleport Connect app provides all the same access to resources as `tsh` and
more, in a friendly graphic user interface. After
[downloading](https://goteleport.com/download/) and installing Teleport Connect,

> you can log in initiate sessions for Server, Database, and Desktop Access all

Reviewer 2: Connect does not support Desktop Access.

Author: Removed.

within a single window.

> 1. Click **CONNECT** to connect to the Teleport plane:

Reviewer 2: ```suggestion
1. Click **CONNECT** to connect to the Teleport cluster:
```

2. Provide the address of your Teleport Cluster (e.g. 
<ScopedBlock scope={["oss", "enterprise"]}>`https://teleport.example.com`</ScopedBlock>
<ScopedBlock scope={["cloud"]}>`https://mytennant.teleport.sh`</ScopedBlock>
) and click **NEXT**.

3. Teleport Connect will ask you for your username, password, and MFA. If this
Teleport plane is integrated with an external Identity Provider (**IdP**), it
may instead prompt you to authenticate in a browser window with that service.

4. Once authenticated, you can browse and connect to all the resources your
user is permitted to access:

   ![An example of Teleport Connect, populated with instances](../img/use-teleport/connect-ui-overview.png)

## Server Access

<Tabs>
<TabItem label="tsh">

`tsh ls` lists the servers you have access to through Server Access (ssh):

```code
$ tsh ls
Node Name           Address         Labels                       
------------------- --------------- ---------------------------- 
server1.example.com 192.0.2.24:3022 access=servers,hostname=server1
server2.example.com 192.0.2.32:3022 access=servers,hostname=server2
```

To connect to a server:

```code

> $tsh ssh server1.example.com

Reviewer 3: It would likely be more useful for end users to have the username included e.g. `tsh ssh ubuntu@server1.example.com`

Author: Adjusted.

```code
$ tsh db ls
Name                                               Description Allowed Users   Labels  Connect
-----------------------------------------------    ----------- --------------- ------- -------------------------
myelastic                                                      [alice elastic] env=dev
mysql-server1 (user: alice, db: teleport_example)              [alice elastic] env=dev tsh db connect mysql-server1
```

To connect to a database server, through `tsh`, you'll need a local client for
that database. For example, to connect to a MySQL or MariaDB database, you'll
need the [MySQL CLI](https://dev.mysql.com/doc/refman/8.0/en/mysql.html) client:

```code
tsh db connect --db-user=alice --db-name=teleport_example mysql-server1 
```

In this example, `teleport_example` is a pre-existing database on the MySQL server:

```sql
SHOW DATABASES;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
| performance_schema |
| teleport_example   |
+--------------------+
4 rows in set (0.09 sec)
```

For other connection types, you can create a local tunnel to point your software to:

```code

> $ tsh proxy db myelastic --db-user=alice --tunnel

Reviewer 3: DB tunneling is not very popular unless it's being used with Teleport Connect or another DB client. If we want to keep these docs concise this section could be removed.

Author: I updated the preceding sentence with what I think makes this section relevant for end-users; access for graphic database clients. I'm no DB admin so I can't say how popular they are versus CLI tools, but my experience creating the Elasticsearch Database Access guide required using a tunnel since using the elastic CLI tool on a system (especially a Mac) without having full Elasticsearch installed locally was a 🤬 and a half.

Keeping in mind that this page is supposed to cover 90% of end users (and not necessarily DB-admins/dev/eng/etc folks, the question is: do we expect the target audience of this page to use GUI database clients?
